### ISSUE TITLE
spine-cpp object files do not link to spine-cocos2dx when compiling using spine runtimes cloned from github on 11/3/2022

### ISSUE THREAD

**Issue by zorro2055** [EsotericSoftware/spine-runtimes#2189]
Hello,

I was actually able to solve this and get the spine-cpp object files to link to spine-cocos2dx by changing a line in spine-cocos2dx/spine-cocos2dx.cmake.

i changed line 28 from:
target_link_libraries(spine-cocos2dx PRIVATE cocos2d)

to:
target_link_libraries(spine-cocos2dx PRIVATE cocos2d spine-cpp)

That fixed the problem.  Should I attempt to push this revision?  (I'll have to figure out how to do that).

**Comment by badlogic** [comment]
If you look at the example project in this repo, you'll find this in the `CmakeList.txt` file of the example:

```
target_link_libraries(${APP_NAME} spine-cpp spine-cocos2dx)
```

That explicitely links both spine-cpp and spine-cococs2dx to the example binary. The `spine-cocos2dx/spine-cocos2dx.cmake` defines both of these targets.

This is also documented in the `README.md` file, see https://github.com/EsotericSoftware/spine-runtimes/tree/4.1/spine-cocos2dx#setup

**Comment by zorro2055** [comment]
Ok, I actually did already have the line in my project Cmakelist.txt file:
target_link_libraries(${APP_NAME} spine-cpp spine-cocos2dx)

In addition, i also have this line with the path updated appropriately in my project Cmakelist.txt:
include(/path/to/spine-runtimes/spine-cocos2dx/spine-cocos2dx.cmake)

To try and get to the bottom of this, I decided to compile the example that is in spine-cocos2dx/example directory.  I did encounter the same error there when line 28 of spine-cocos2dx/spine-cocos2dx.cmake is:
target_link_libraries(spine-cocos2dx PRIVATE cocos2d)

When I changed that line to:
target_link_libraries(spine-cocos2dx PRIVATE cocos2d spine-cpp)
it compiled fine and executed successfully.

I am guessing that the line in the project file Cmakelist.txt:
target_link_libraries(${APP_NAME} spine-cpp spine-cocos2dx)
only tells cmake to link spine-cpp to $(APP_NAME), but does NOT tell spine-cpp to link to spine-cocos2dx (which appears to need it).

Am I missing something?  I am compiling in Gentoo Linux at the command line.  Partial compiler and executable (if one was created) outputs below.

Partial compiler output when line in spine-cocos2dx is "target_link_libraries(spine-cocos2dx PRIVATE cocos2d)":
[100%] Linking CXX executable bin/spine-cocos2dx-example/spine-cocos2dx-example
/usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../x86_64-pc-linux-gnu/bin/ld: libspine-cocos2dx.a(SkeletonRenderer.cpp.o): in function `spine::SkeletonRenderer::initialize()':
SkeletonRenderer.cpp:(.text+0x241): undefined reference to `spine::SkeletonClipping::SkeletonClipping()'
/usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../x86_64-pc-linux-gnu/bin/ld: libspine-cocos2dx.a(SkeletonRenderer.cpp.o): in function `spine::SkeletonRenderer::initWithBinaryFile(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, spine::Atlas*, float)':
SkeletonRenderer.cpp:(.text+0x1306): undefined reference to `spine::SkeletonBinary::SkeletonBinary(spine::AttachmentLoader*, bool)'
/usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../x86_64-pc-linux-gnu/bin/ld: SkeletonRenderer.cpp:(.text+0x135b): undefined reference to `spine::SkeletonBinary::readSkeletonDataFile(spine::String const&)'
/usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../x86_64-pc-linux-gnu/bin/ld: SkeletonRenderer.cpp:(.text+0x13c8): undefined reference to `spine::SkeletonBinary::~SkeletonBinary()'
/usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../x86_64-pc-linux-gnu/bin/ld: SkeletonRenderer.cpp:(.text+0x1426): undefined reference to `spine::SkeletonBinary::~SkeletonBinary()'
/usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../x86_64-pc-linux-gnu/bin/ld: libspine-cocos2dx.a(SkeletonRenderer.cpp.o): in function `spine::SkeletonRenderer::initWithBinaryFile(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, float)':
SkeletonRenderer.cpp:(.text+0x1568): undefined reference to `spine::SkeletonBinary::SkeletonBinary(spine::AttachmentLoader*, bool)'
/usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../x86_64-pc-linux-gnu/bin/ld: SkeletonRenderer.cpp:(.text+0x15bd): undefined reference to `spine::SkeletonBinary::readSkeletonDataFile(spine::String const&)'
/usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../x86_64-pc-linux-gnu/bin/ld: SkeletonRenderer.cpp:(.text+0x1638): undefined reference to `spine::SkeletonBinary::~SkeletonBinary()'
/usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../x86_64-pc-linux-gnu/bin/ld: SkeletonRenderer.cpp:(.text+0x16d8): undefined reference to `spine::SkeletonBinary::~SkeletonBinary()'
/usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../x86_64-pc-linux-gnu/bin/ld: libspine-cocos2dx.a(SkeletonRenderer.cpp.o): in function `spine::SkeletonRenderer::draw(cocos2d::Renderer*, cocos2d::Mat4 const&, unsigned int)':
SkeletonRenderer.cpp:(.text+0x1b90): undefined reference to `spine::SkeletonClipping::clipEnd(spine::Slot&)'
/usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../x86_64-pc-linux-gnu/bin/ld: SkeletonRenderer.cpp:(.text+0x2348): undefined reference to `spine::SkeletonClipping::clipStart(spine::Slot&, spine::ClippingAttachment*)'
/usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../x86_64-pc-linux-gnu/bin/ld: SkeletonRenderer.cpp:(.text+0x236d): undefined reference to `spine::SkeletonClipping::clipEnd(spine::Slot&)'
/usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../x86_64-pc-linux-gnu/bin/ld: SkeletonRenderer.cpp:(.text+0x2479): undefined reference to `spine::SkeletonClipping::clipEnd(spine::Slot&)'
/usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../x86_64-pc-linux-gnu/bin/ld: SkeletonRenderer.cpp:(.text+0x26c9): undefined reference to `spine::SkeletonClipping::isClipping()'
/usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../x86_64-pc-linux-gnu/bin/ld: SkeletonRenderer.cpp:(.text+0x270f): undefined reference to `spine::SkeletonClipping::clipTriangles(float*, unsigned short*, unsigned long, float*, unsigned long)'
/usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../x86_64-pc-linux-gnu/bin/ld: SkeletonRenderer.cpp:(.text+0x2739): undefined reference to `spine::SkeletonClipping::getClippedTriangles()'
/usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../x86_64-pc-linux-gnu/bin/ld: SkeletonRenderer.cpp:(.text+0x276b): undefined reference to `spine::SkeletonClipping::clipEnd(spine::Slot&)'
/usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../x86_64-pc-linux-gnu/bin/ld: SkeletonRenderer.cpp:(.text+0x2786): undefined reference to `spine::SkeletonClipping::getClippedVertices()'
/usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../x86_64-pc-linux-gnu/bin/ld: SkeletonRenderer.cpp:(.text+0x27c2): undefined reference to `spine::SkeletonClipping::getClippedTriangles()'
/usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../x86_64-pc-linux-gnu/bin/ld: SkeletonRenderer.cpp:(.text+0x27fb): undefined reference to `spine::SkeletonClipping::getClippedTriangles()'
/usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../x86_64-pc-linux-gnu/bin/ld: SkeletonRenderer.cpp:(.text+0x281d): undefined reference to `spine::SkeletonClipping::getClippedTriangles()'
/usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../x86_64-pc-linux-gnu/bin/ld: SkeletonRenderer.cpp:(.text+0x28b4): undefined reference to `spine::SkeletonClipping::getClippedVertices()'
/usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../x86_64-pc-linux-gnu/bin/ld: SkeletonRenderer.cpp:(.text+0x28d9): undefined reference to `spine::SkeletonClipping::getClippedUVs()'
/usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../x86_64-pc-linux-gnu/bin/ld: SkeletonRenderer.cpp:(.text+0x2b07): undefined reference to `spine::SkeletonClipping::isClipping()'
/usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../x86_64-pc-linux-gnu/bin/ld: SkeletonRenderer.cpp:(.text+0x2b44): undefined reference to `spine::SkeletonClipping::clipTriangles(float*, unsigned short*, unsigned long, float*, unsigned long)'
/usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../x86_64-pc-linux-gnu/bin/ld: SkeletonRenderer.cpp:(.text+0x2b70): undefined reference to `spine::SkeletonClipping::getClippedTriangles()'
/usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../x86_64-pc-linux-gnu/bin/ld: SkeletonRenderer.cpp:(.text+0x2ba2): undefined reference to `spine::SkeletonClipping::clipEnd(spine::Slot&)'
/usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../x86_64-pc-linux-gnu/bin/ld: SkeletonRenderer.cpp:(.text+0x2bbd): undefined reference to `spine::SkeletonClipping::getClippedVertices()'
/usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../x86_64-pc-linux-gnu/bin/ld: SkeletonRenderer.cpp:(.text+0x2bfb): undefined reference to `spine::SkeletonClipping::getClippedTriangles()'
/usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../x86_64-pc-linux-gnu/bin/ld: SkeletonRenderer.cpp:(.text+0x2c36): undefined reference to `spine::SkeletonClipping::getClippedTriangles()'
/usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../x86_64-pc-linux-gnu/bin/ld: SkeletonRenderer.cpp:(.text+0x2c58): undefined reference to `spine::SkeletonClipping::getClippedTriangles()'
/usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../x86_64-pc-linux-gnu/bin/ld: SkeletonRenderer.cpp:(.text+0x2cfd): undefined reference to `spine::SkeletonClipping::getClippedVertices()'
/usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../x86_64-pc-linux-gnu/bin/ld: SkeletonRenderer.cpp:(.text+0x2d22): undefined reference to `spine::SkeletonClipping::getClippedUVs()'
/usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../x86_64-pc-linux-gnu/bin/ld: SkeletonRenderer.cpp:(.text+0x2f83): undefined reference to `spine::SkeletonClipping::clipEnd(spine::Slot&)'
/usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../x86_64-pc-linux-gnu/bin/ld: SkeletonRenderer.cpp:(.text+0x2fb2): undefined reference to `spine::SkeletonClipping::clipEnd()'
/usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../x86_64-pc-linux-gnu/bin/ld: libspine-cocos2dx.a(spine-cocos2dx.cpp.o): in function `spine::Cocos2dTextureLoader::Cocos2dTextureLoader()':
spine-cocos2dx.cpp:(.text+0x12c): undefined reference to `spine::TextureLoader::TextureLoader()'
/usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../x86_64-pc-linux-gnu/bin/ld: libspine-cocos2dx.a(spine-cocos2dx.cpp.o): in function `spine::Cocos2dTextureLoader::~Cocos2dTextureLoader()':
spine-cocos2dx.cpp:(.text+0x164): undefined reference to `spine::TextureLoader::~TextureLoader()'
/usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../x86_64-pc-linux-gnu/bin/ld: libspine-cocos2dx.a(spine-cocos2dx.cpp.o):(.data.rel.ro._ZTIN5spine20Cocos2dTextureLoaderE[_ZTIN5spine20Cocos2dTextureLoaderE]+0x10): undefined reference to `typeinfo for spine::TextureLoader'
collect2: error: ld returned 1 exit status
make[2]: *** [CMakeFiles/spine-cocos2dx-example.dir/build.make:313: bin/spine-cocos2dx-example/spine-cocos2dx-example] Error 1
make[1]: *** [CMakeFiles/Makefile2:577: CMakeFiles/spine-cocos2dx-example.dir/all] Error 2
make: *** [Makefile:91: all] Error 2
johnd@Airwolf ~/Documents/SmallGameCocos/spine-runtimes/spine-cocos2dx/example/linux-build $


Partial compiler and execution output when line in spine-cocos2dx is "target_link_libraries(spine-cocos2dx PRIVATE cocos2d spine-cpp)":
[100%] Linking CXX executable bin/spine-cocos2dx-example/spine-cocos2dx-example
[100%] Built target spine-cocos2dx-example
johnd@Airwolf ~/Documents/SmallGameCocos/spine-runtimes/spine-cocos2dx/example/linux-build $ bin/spine-cocos2dx-example/spine-cocos2dx-example 
Ready for GLSL
Ready for OpenGL 2.0
/home/johnd/Documents/SmallGameCocos/spine-runtimes/spine-cpp/spine-cpp/src/spine/Atlas.cpp:59 (address 0x55df29d7d2c0): Double free or not allocated through SpineExtension
/home/johnd/Documents/SmallGameCocos/spine-runtimes/spine-cpp/spine-cpp/src/spine/SkeletonBinary.cpp:359 (address 0x55df29d971c0): Double free or not allocated through SpineExtension

You can see there are some warnings while running, but it worked fine.  The example executable shows a dragon like monster flying.

**Comment by badlogic** [comment]
How do you invoke CMake for our example project?

**Comment by zorro2055** [comment]
cmake ..

from directory:
~/Documents/SmallGameCocos/spine-runtimes/spine-cocos2dx/example/linux-build

where I created the empty linux-build directory.

**Comment by zorro2055** [comment]
Below is the cmake version I am using too if that helps?

johnd@Airwolf ~/Documents/SmallGameCocos/spine-runtimes/spine-cocos2dx/example/linux-build $ cmake --version
cmake version 3.24.1

CMake suite maintained and supported by Kitware (kitware.com/cmake).

**Comment by badlogic** [comment]
Thanks for the info. All of that looks g9od to me. I'll try on a Linux
machine on Monday.

**Comment by zorro2055** [comment]
Ok, thanks, much appreciated!

**Comment by badlogic** [comment]
Sorry it took a bit longer! I've now reproed this on Ubuntu 22.04, while previously I was testing on Ubuntu 16. I'm not entirely sure what has changed, as the CMake versions in both cases are the same. That said, explicitely having spine-cocos2d link to spine-cpp in `spine-cocos2dx.cmake` solves this issue. Weirdly enough, IDE project generation for Visual Studio and Xcode are not affected by this, neither is generating the Android project files.

Thanks for reporting! This is now fixed on the 4.1 and 4.2-beta branches.
